### ISSUE TITLE
s3: remote execution doesn't use provided credential parameters

### ISSUE THREAD

**Issue by lorrin** [ansible/ansible#5211]
I was trying to use the s3 module to have the target host directly fetch artifacts from S3. It failed with an authentication error despite the fact that I had provided credentials (via `aws_secret_key` and `aws_secret_access_key` parameters). I worked around by deploying a `~/.boto` to the host.

Play:

```
- name: debug
  debug: msg="key {{aws_access_key_id}} secret {{aws_secret_access_key}}"

- name: Fetch Foo
  s3: bucket=foo_bucket object=/foo/bar/foobar.tgz dest=/tmp/foobar.tgz mode=get aws_access_key={{aws_access_key_id}} aws_secret_key={{aws_secret_access_key}}
```

Failure:

```
failed: [ec2-50-16-55-243.compute-1.amazonaws.com] => {"failed": true, "item": ""}
msg: No handler was ready to authenticate. 1 handlers were checked. ['HmacAuthV1Handler'] Check your credentials
```

This is similar to https://github.com/ansible/ansible/issues/4212. In that issue, the user had set `AWS_SECRET_ACCESS_KEY` and `AWS_ACCESS_KEY_ID` locally and the operation failed because those shell environment variables weren't present on the target machine. That seems reasonable. Maybe something similar is going on here? But it seems to me that in this case these parameters should be forwarded over the wire.

With `-vvvv` it looks like this (after some redaction):

```
<ec2-54-226-182-229.compute-1.amazonaws.com> ESTABLISH CONNECTION FOR USER: ubuntu
<ec2-54-226-182-229.compute-1.amazonaws.com> EXEC ['ssh', '-tt', '-vvv', '-o', 'ControlMaster=auto', '-o', 'ControlPersist=60s', '-o', 'ControlPath=/Users/lhn/.ansible/cp/ansible-ssh-%h-%p-%r', '-o', 'StrictHostKeyChecking=no', '-o', 'Port=22', '-o', 'IdentityFile=./mykey.pem', '-o', 'KbdInteractiveAuthentication=no', '-o', 'PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey', '-o', 'PasswordAuthentication=no', '-o', 'User=ubuntu', '-o', 'ConnectTimeout=10', u'ec2-54-226-182-229.compute-1.amazonaws.com', "/bin/sh -c 'mkdir -p $HOME/.ansible/tmp/ansible-1386375985.93-120892534651087 && chmod a+rx $HOME/.ansible/tmp/ansible-1386375985.93-120892534651087 && echo $HOME/.ansible/tmp/ansible-1386375985.93-120892534651087'"]
<ec2-54-226-182-229.compute-1.amazonaws.com> REMOTE_MODULE s3 bucket=foo_bucket object=/foo/bar/foobar.tgz dest=/tmp/foobar.tgz mode=get aws_access_key=**** aws_secret_key=****
<ec2-54-226-182-229.compute-1.amazonaws.com> PUT /var/folders/52/td9wml9110q0zc7s76_ckz6r0000gv/T/tmp5_foHI TO /home/ubuntu/.ansible/tmp/ansible-1386375985.93-120892534651087/s3
<ec2-54-226-182-229.compute-1.amazonaws.com> PUT /var/folders/52/td9wml9110q0zc7s76_ckz6r0000gv/T/tmpGjXDub TO /home/ubuntu/.ansible/tmp/ansible-1386375985.93-120892534651087/arguments
<ec2-54-226-182-229.compute-1.amazonaws.com> EXEC ['ssh', '-tt', '-vvv', '-o', 'ControlMaster=auto', '-o', 'ControlPersist=60s', '-o', 'ControlPath=/Users/lhn/.ansible/cp/ansible-ssh-%h-%p-%r', '-o', 'StrictHostKeyChecking=no', '-o', 'Port=22', '-o', 'IdentityFile=./mykey.pem', '-o', 'KbdInteractiveAuthentication=no', '-o', 'PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey', '-o', 'PasswordAuthentication=no', '-o', 'User=ubuntu', '-o', 'ConnectTimeout=10', u'ec2-54-226-182-229.compute-1.amazonaws.com', "/bin/sh -c '/usr/bin/python /home/ubuntu/.ansible/tmp/ansible-1386375985.93-120892534651087/s3 /home/ubuntu/.ansible/tmp/ansible-1386375985.93-120892534651087/arguments; rm -rf /home/ubuntu/.ansible/tmp/ansible-1386375985.93-120892534651087/ >/dev/null 2>&1'"]
failed: [ec2-54-226-182-229.compute-1.amazonaws.com] => {"failed": true, "item": ""}
msg: No handler was ready to authenticate. 1 handlers were checked. ['HmacAuthV1Handler'] Check your credentials
```

Is there a way to make it log the contents of the temp files being executed?

**Comment by jctanner** [comment]
@lorrin -vvvv prints out the remote script path and execution method ...

```
/usr/bin/python /home/ubuntu/.ansible/tmp/ansible-1386375985.93-120892534651087/s3 /home/ubuntu/.ansible/tmp/ansible-1386375985.93-120892534651087/arguments
```

What you want is to make sure it doesn't delete the file so that you can go test it. To do that:

```
ANSIBLE_KEEP_REMOTE_FILES=1 ansible-playbook <args>
```

**Comment by mpdehaan** [comment]
Another way to test modules free of Ansible (and with debug), BTW, is to use the ./hacking/test-module script

./hacking/test-module -m cloud/s3 -a "arguments=go right=here"

This requires an ansible checkout on the remote machine, but does not require that things be installed.

You can source ./hacking/env-setup if you want beforehand to activate the environment.

The module common code used by all AWS modules should load this consistently for s3, which in the code is handled by:

```
ec2_url, aws_access_key, aws_secret_key, region = get_ec2_creds(module)
```

So possibly something else is in fact going on.

**Comment by jctanner** [comment]
@lorrin just re-read through your initial description. It seems as though you have environment variables on your control host for the ec2 credentials, and are expecting ansible to forward that environment over to the client machine. That is not quite how it works.

In most examples for using EC2,  environment vars are set on the control host and then the call to the ec2 module is a "local" play which would have access to those vars. If the the ec2 environment vars are not set on a client node and you then call the ec2 module on that client, it's only going to have the default environment variables for that remote user. Ansible will not forward them from the control host.

There is a way to further define the environment for a task:
http://www.ansibleworks.com/docs/playbooks_environment.html

You can also lookup environment vars and then register those as facts for later plays:
http://www.ansibleworks.com/docs/faq.html#how-do-i-access-shell-environment-variables
http://www.ansibleworks.com/docs/modules.html#set-fact

**Comment by lorrin** [comment]
@jctanner , that's not quite what I'm expecting. I get that environment variables don't forward. I'm passing the credentials as arguments to the module invocation in the play:

```
s3: bucket=foo_bucket object=/foo/bar/foobar.tgz dest=/tmp/foobar.tgz mode=get aws_access_key={{aws_access_key_id}} aws_secret_key={{aws_secret_access_key}}
```

It seems very unintuitive to me that those would not make it over the wire. But if that is the indented behavior, I think http://www.ansibleworks.com/docs/modules.html#s3 should call out that the parameters don't work in the general case and can only be used when operating on the control host. (Or even remove them as a mis-feature?)

**Comment by jctanner** [comment]
@lorrin were you able to validate that the MODULE_ARGS line in the remote python script actually contained the right credentials?

**Comment by lorrin** [comment]
Just tried it. They _do_ get set in `~/.ansible/.../arguments`!

```
bucket=my_bucket object=/foobar.tgz mode=get aws_access_key=FOO aws_secret_key=BAR
```

I think this is the bug (snippet from the `s3` script deployed alongside the `arguments` file):

```
def get_ec2_creds(module):

    # Check module args for credentials, then check environment vars

    ec2_url = module.params.get('ec2_url')
    ec2_secret_key = module.params.get('ec2_secret_key')
    ec2_access_key = module.params.get('ec2_access_key')
    region = module.params.get('region')
```

Note "ec2" rather than "aws" in the key name.

**Comment by willthames** [comment]
I've just managed to successfully get this all to work without issue. 

I created an instance in EC2 to test this with (doesn't need to be in EC2, just separate to my control node) and then ran the following with s3creds.yml containing secret key, access key, instance etc. 

```
- hosts: "{{instance}}"
  user: "ec2-user"
  vars_files:
  - s3creds.yml

  tasks:
  - name: debug
    debug: msg="key {{aws_access_key}} secret {{aws_secret_key}}"

  - name: Fetch Foo
    s3: bucket="{{bucket}}" object=handypython dest=/tmp/handypython mode=get aws_access_key={{aws_access_key}} aws_secret_key={{aws_secret_key}}
```

**Comment by willthames** [comment]
Because of how `AnsibleModule` works, `modules.params` has `aws_secret_key` and `aws_access_key` setup as copies of `ec2_secret_key` and `ec2_access_key` (this is done in `_handle_aliases`).

If it's not working for you, and you have `ANSIBLE_KEEP_REMOTE_FILES` set to 1, just stick 

```
import pdb
pdb.set_trace()
```

in the resulting s3 file and trace the program through and see what is not happening as it should be.

**Comment by lorrin** [comment]
Hmm. When I trace through, `aws_access_key` and `aws_secret_key` are `None`. I'm suspicious that the module.params is not set up right. It's like this:

```
    module = AnsibleModule(
        argument_spec = dict(
            bucket         = dict(required=True),
            object         = dict(),
            src            = dict(),
            dest           = dict(default=None),
            mode           = dict(choices=['get', 'put', 'delete', 'create', 'geturl', 'getstr'], required=True),
            expiry         = dict(default=600, aliases=['expiration']),
            s3_url         = dict(aliases=['S3_URL']),
            aws_secret_key  = dict(aliases=['ec2_secret_key', 'secret_key'], no_log=True, required=False),
            aws_access_key  = dict(aliases=['ec2_access_key', 'access_key'], required=False),
            overwrite      = dict(aliases=['force'], default=True, type='bool'),
        ),
    )
```

That looks to me like it defines how an 'aws_secret_key' and 'aws_access_key' argument are configured (with various aliases). (Looks right!) But then:

```
def get_ec2_creds(module):

    # Check module args for credentials, then check environment vars

    ec2_url = module.params.get('ec2_url')
    ec2_secret_key = module.params.get('ec2_secret_key')
    ec2_access_key = module.params.get('ec2_access_key')
    region = module.params.get('region')
```

That looks wrong -- doesn't it need to use `'aws_secret_key'` and `'aws_access_key'` since those are how the module params are configured? I would have thought `'ec2_secret_key'` was just an alternate way the argument could be supplied by the user, not an alternate that could be used within the code.

To wit:

```
> /home/ubuntu/.ansible/tmp/ansible-1386712687.2-214041998230334/s3(1488)get_ec2_creds()
-> ec2_secret_key = module.params.get('ec2_secret_key')
(Pdb) s
> /home/ubuntu/.ansible/tmp/ansible-1386712687.2-214041998230334/s3(1489)get_ec2_creds()
-> ec2_access_key = module.params.get('ec2_access_key')
(Pdb) p ec2_secret_key
None
(Pdb) p module.params.get('aws_secret_key')
'my_secret_key_showed_up_here'
(Pdb)
```

**Comment by willthames** [comment]
That's not the latest version of the `s3` module. What version are you running?

https://github.com/ansible/ansible/blob/devel/library/cloud/s3#L261-L262

**Comment by willthames** [comment]
Looks like this was fixed _two weeks_ ago in commit 26374d89a

**Comment by lorrin** [comment]
Hey, that's the fix I thought was needed! :-) I assumed I was current, but I'm not -- still on 1.4. I'll go grab 1.4.1. Thanks for helping me track it down.
